x/bank/keeper/keeper.go

```diff
@@ -398,28 +373,40 @@ func (k BaseKeeper) BurnCoins(ctx sdk.Context, moduleName string, amt sdk.Coins)
 		panic(sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "module account %s does not have permissions to burn tokens", moduleName))
 	}
 
-	err := k.subUnlockedCoins(ctx, acc.GetAddress(), amt)
+	err := k.subUnlockedCoins(ctx, acc.GetAddress(), amounts)
 	if err != nil {
 		return err
 	}
 
-	// update total supply
-	supply := k.GetTotalSupply(ctx)
-	supply = supply.Sub(amt)
-
-	k.setSupply(ctx, supply)
+	var newSupply sdk.Coins
+	for _, amount := range amounts {
+		supply := k.GetSupply(ctx, amount.GetDenom())
+		supply = supply.Sub(amount)
+		newSupply = append(newSupply, supply)
+	}
+	k.setSupply(ctx, newSupply)
```

Comment:
I don't see a need to bundle this into Coins at all. Instead I think we should just be setting supply by denom in the above for loop. The whole Coins abstraction is just unnecessary here IMHO.

Reply:
+1 for inlining. The function is used only 2 times, so not a big trade off.